### PR TITLE
!!! BUGFIX: Always unset node property, if set to null

### LINKED ISSUE #4467: 9.0 Node property is set to null instead of unset in database
related https://github.com/neos/neos-development-collection/issues/4304

this comment says that we treat null as unset, but i found out that this is not entirely true ;)
https://github.com/neos/neos-development-collection/blob/782b9585bb53f71a326c99c9fbf85671418f91ef/Neos.ContentRepository.Core/Classes/Feature/NodeModification/Event/NodePropertiesWereSet.php#L32-L33

creating a node and setting in the initial properties a defined property `foo` to `null` will show up in the CR explicitly.

### PR DESCRIPTION
**Pull request by mhsdesign**
Previously we made a distinction between unsetting a node property and setting it to null on the write side.

There was setting to null `SerializedPropertyValues::fromArray(['tagName' => new SerializedPropertyValue(null, 'string')])` and unset `SerializedPropertyValues::fromArray(['tagName' => null])`

Unsetting a property would only be possible via php api by creating a custom `SerializedPropertyValues`.

The read site treats it the same and returns true in hasProperty only if there is a non null value in the database.

As discussed with @bwaidelich 

>  I would recommend to **not make a difference whether a property wasn't set initially or whether it was removed lateron**.
While it might be useful to expose this information at some point, we should do so _explicitly_ and not via the `null`-behavior of some adapter.

Previously if a property tagName was set to `null`

the `NodePropertiesWereSet` event would contain
```
"propertyValues":{"tagName":{"value":null,"type":"string"}}
```
and the properties of the node in the projection as well
```
"tagName":{"value":null,"type":"string"}
```

-----

which this change the the `NodePropertiesWereSet` event would contain:
```
"propertyValues":{"tagName":null}
```
and the properties of the node in the projection would show no hint of this property.

-------

Our logic to decide between unsetting properties and setting them to null was spread through a lot of classes and is a sheer wonder that it worked :D Now with the instruction of the `UnsetPropertyValue`, it is much more obvious how the behaviour works. 

solves: #4467

> [!IMPORTANT]  
> This change is breaking to your existing content repository data. To migrate your events please execute:
> ```sh
> flow migrateevents:migratePropertiesToUnset
> flow cr:projectionReplay contentGraph
> ```
> otherwise you will face this error `SerializedPropertyValue::__construct(): Argument #1 ($value) must be of type ArrayObject|array|string|int|float|bool, null given`

### CODE REVIEW

**Comment by bwaidelich** [comment]
From the description and the code itself, I don't really understand this change. But I have the strong feeling that is should be backed by a test

**Comment by mhsdesign** [comment]
Yes i see, sorry for being so spare about the details.

Id like to "quickly" chat with you about this to structure my mind - after that i could probably write it down more easily, and we could also discuss if my expectations are correct.

**Comment by bwaidelich** [comment]
> Id like to "quickly" chat with you about this to structure my mind

Those quotes around "quickly" make me suspicious :)
No, sure thing, let's talk about this sometimes next week!

**Comment by bwaidelich** [comment]
I added some thoughts to https://github.com/neos/neos-development-collection/issues/4304#issuecomment-1604042784

**Review by bwaidelich — DISMISSED** [review]

**Review by skurfuerst — APPROVED** [review]
looks good :) wonder why the stuff before was done ;)

**Comment by mhsdesign** [comment]
Now i only have to come up with a test for this ;) (Thats why its WIP - and also need to rebase it)

File: Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php [hunk 1]
```diff
@@ -44,8 +44,8 @@ public function serializePropertyValues(
         $serializedPropertyValues = [];
 
         foreach ($propertyValuesToWrite->values as $propertyName => $propertyValue) {
-            if (!isset($nodeType->getProperties()[$propertyName]) && $propertyValue === null) {
-                // The property is undefined and the value null => we want to remove it, so we set it to null
+            if ($propertyValue === null) {
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php]: `!isset($nodeType->getProperties()[$propertyName])` is at that point always `false` because we only allow to setProperties that actually exist in the current schema ([validateProperties](https://github.com/neos/neos-development-collection/blob/9c8f3077b337ae38556d43cbd75142ee1484944a/Neos.ContentRepository.Core/Classes/Feature/NodeCreation/NodeCreation.php#L117-L130)) - thats why the the code path was previously never taken.

**Comment by mhsdesign** [comment]
As we discussed previously it is an implementation detail of the adapter to save null values to the database or not, thats why its a bit hard to come up with a test? And as everything passes that seems proof enough this refactoring is correct?

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php [hunk 2]
```diff
@@ -18,16 +18,17 @@
...
  */
 final class SerializedPropertyValue implements \JsonSerializable
 {
     /**
-     * @param int|float|string|bool|array<int|string,mixed>|\ArrayObject<int|string,mixed>|null $value
+     * @param int|float|string|bool|array<int|string,mixed>|\ArrayObject<int|string,mixed> $value
      */
     public function __construct(
-        public readonly int|float|string|bool|array|\ArrayObject|null $value,
+        public readonly int|float|string|bool|array|\ArrayObject $value,
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php:35]: its _might_ be uncool to have this signature and always having to check the value before against null.
We could also have a static factory that returns either `SerializedPropertyValue` or `UnsetPropertyValue` if null was accidentally still passed?

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValues.php [hunk 3]
```diff
@@ -52,10 +49,14 @@ public static function fromArray(array $propertyValues): self
...
-                // this case means we want to un-set a property.
-                $values[$propertyName] = null;
+            if ($propertyValue === null || $propertyValue instanceof UnsetPropertyValue) {
+                $values[$propertyName] = UnsetPropertyValue::get();
             } elseif (is_array($propertyValue)) {
+                if (array_key_exists('value', $propertyValue) && $propertyValue['value'] === null) {
+                    // legacy
+                    $values[$propertyName] = UnsetPropertyValue::get();
+                    continue;
+                }
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValues.php]: without this upcasting, a replay on older event would lead to:

> SerializedPropertyValue::__construct(): Argument #1 ($value) must be of type ArrayObject|array|string|int|float|bool, null given, called in 

alternatively, we could possibly migrate the events table?

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php:35]: solved via `SerializedPropertyValue::create`

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValues.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValues.php]: solved via `SerializedPropertyValue::create`

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php [hunk 4]
```diff
@@ -18,24 +18,40 @@
...
+     */
+    public static function create(
+        int|float|string|bool|array|\ArrayObject|null $value,
+        string $type
+    ): self|UnsetPropertyValue {
+        if ($value === null) {
+            return UnsetPropertyValue::get();
+        }
+        return new self($value, $type);
+    }
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php]: Do we need to expose the "null"-hack to the new constructor?
I would suggest to change this to:
```suggestion
    public static function create(
        int|float|string|bool|array|\ArrayObject $value,
        string $type
    ): self|UnsetPropertyValue {
        return new self($value, $type);
    }
```

and then...

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php [hunk 5]
```diff
@@ -44,7 +60,7 @@ public static function fromArray(array $valueAndType): self
             throw new \InvalidArgumentException('Missing array key "type"', 1546524609);
         }
 
-        return new self($valueAndType['value'], $valueAndType['type']);
+        return self::create($valueAndType['value'], $valueAndType['type']);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php]: ...adjust this accordingly:
```suggestion
        return $valueAndType['value'] === null ? UnsetPropertyValue::get() : self::create($valueAndType['value'], $valueAndType['type']);
```

**Review by bwaidelich — DISMISSED** [review]
Sorry for being complicated, but I dislike that one of the highest-level APIs now has this awkward `withoutUnsets()` method.
I am probably missing something but I expected `SerializedProperties` to be just that, a set of serialized properties. But that they would never contain any `null`-values.
I.e. that we would filter them out at _write_ time.

I know, that we need it to _unset_ previously existing properties, but I thought that we could put that logic into `PropertyValuesToWrite` rather than having something like `Node->properties->serialized->withoutUnsets()` which does not make sense to me on the read side at all

**Comment by mhsdesign** [comment]
We discussed to introduce a `PropertyNames` dto and a new property in the `SetSerializedProperties` command / event `$propertiesToUnset`.

Additionally we briefly discussed that a property converter for the properties should probably never return `null`, as this would cause confusion about `hasProperty` seeminly not working.

**Comment by mhsdesign** [comment]
So i also just wrote a database migration that will update the events. Adding a php compatibility layer would just permanently introduce complexity and possibly bugs.

The following things have to be migrated:

- `NodePropertiesWereSet` `payload`
- `SetSerializedNodeProperties` in `metadata`
example:
```
"propertyValues":{"tagName":{"value":null,"type":"string"}}
->
"propertyValues":[],"propertiesToUnset":["tagName"]
```

Here we just have to omit the null values because its a new node either way:

- `NodeAggregateWithNodeWasCreated` `payload`
- `CreateNodeAggregateWithNodeAndSerializedProperties` in `metadata`
```
"initialPropertyValues":{"title":{"value":"Blog","type":"string"},"titleOverride":{"value":null,"type":"string"},"uriPathSegment":{"value":"blog","type":"string"}}
->
"initialPropertyValues":{"title":{"value":"Blog","type":"string"},"uriPathSegment":{"value":"blog","type":"string"}}
```
- `CopyNodesRecursively` in `metadata`

```
// somewhere deep recursive in $nodeTreeToInsert
"propertyValues":{"senderName":{"value":null,"type":"string"}}}
-> 
"propertyValues":{}
```



Also the basically impossible `"initialPropertyValues":{"tagName":null}` and `"propertyValues":{"tagName":null` will be migrated.

**Comment by mhsdesign** [comment]
This is now ready to be reviewed :D And with the migration mergeable.

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyNames.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+namespace Neos\ContentRepository\Core\Feature\NodeModification\Dto;
+
+use Neos\ContentRepository\Core\NodeType\NodeType;
+use Neos\ContentRepository\Core\SharedModel\Node\PropertyName;
+
+/**
+ * Todo move to shared and use {@see PropertyName} ???
+ * @api used as part of commands/events
+ */
+final readonly class PropertyNames implements \JsonSerializable
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyNames.php]: Todo move to shared and use {@see PropertyName} ???

**Comment by ahaeslich** [comment]
Just tested the migration on a project with ~100k events. 38 events were migrated, for example:

Before: `"propertyValues":{"image":{"value":null,"type":"Neos\\Media\\Domain\\Model\\ImageInterface"}}`
After: `"propertyValues":[],"propertiesToUnset":["image"]`

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        $backupEventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId)
+            . '_bak_' . date('Y_m_d_H_i_s');
+        $outputFn(sprintf('Backup: copying events table to %s', $backupEventTableName));
+
+        $this->copyEventTable($backupEventTableName);
+
+        $streamName = VirtualStreamName::all();
+        $eventStream = $this->eventStore->load($streamName);
+        foreach ($eventStream as $eventEnvelope) {
+            $outputRewriteNotice = fn(string $message) => $outputFn(sprintf('%s@%s %s', $eventEnvelope->event->type->value, $eventEnvelope->sequenceNumber->value, $message));
```
Comment (ahaeslich) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: Maybe we can reduce the output. If one has many events, warnings or migration messages are hard to find 😅. This also applies to line 110

Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: yes i wasnt aware of 100k nodes :O Sorry to your terminal

**Comment by mhsdesign** [comment]
Okay with @nezaniel help i also tested the migration on his starship's event stream (of 30K events) where the migration could successfully migrate `12884` events.

Excerpt:

```
2292@NodeAggregateWithNodeWasCreated Payload: Removed 1 $initialPropertyValues
2301@NodeAggregateWithNodeWasCreated Payload: Removed 5 $initialPropertyValues
2350@NodeAggregateWithNodeWasCreated Payload: Removed 1 $initialPropertyValues
...
77513@NodePropertiesWereSet Payload: Migrated 1 $unsetProperties
77513@NodePropertiesWereSet Metadata: Migrated 1 $unsetProperties
77514@NodePropertiesWereSet Payload: Migrated 1 $unsetProperties
77514@NodePropertiesWereSet Metadata: Migrated 1 $unsetProperties
```

I think having the migration now tested with 100K events, 30K and with the demo, we can be sure that it works.

I could also imagine diffing the content graph node projections for full insurance, but i fear that would be to much data :D

File: Neos.ContentRepository.BehavioralTests/Classes/PhpstanRules/InternalMethodsNotAllowedOutsideContentRepositoryRule.php [hunk 8]
```diff
@@ -40,6 +40,8 @@ public function processNode(Node $node, Scope $scope): array
                 || str_starts_with($scope->getNamespace(), 'Neos\ContentRepository\Export')
                 || str_starts_with($scope->getNamespace(), 'Neos\ContentRepository\LegacyNodeMigration')
                 || str_starts_with($scope->getNamespace(), 'Neos\ContentRepository\StructureAdjustment')
+                // we do some next level stuff in the migrations, and thus we do illegal stuff:
```
Comment (kitsunet) [inline Neos.ContentRepository.BehavioralTests/Classes/PhpstanRules/InternalMethodsNotAllowedOutsideContentRepositoryRule.php]: can we maybe describe this more matter of factly, because this comment doesn't help me at all to understand what's going on here.

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ContentStreamForking/ForkContentStreamWithoutDimensions.feature [hunk 9]
```diff
@@ -46,6 +46,7 @@ Feature: ForkContentStream Without Dimensions
       | originDimensionSpacePoint    | {}                                                      |
       | affectedDimensionSpacePoints | [{}]                                                    |
       | propertyValues               | {"text": {"value": "original value", "type": "string"}} |
+      | propertiesToUnset            | {}                                                      |
```
Comment (kitsunet) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ContentStreamForking/ForkContentStreamWithoutDimensions.feature:49]: This could default to empty so you don't need to add it to every feature?

File: Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php [hunk 10]
```diff
@@ -72,26 +67,35 @@ private function serializePropertyValue(
...
+            $serializedPropertyValue = $this->serializer->normalize($propertyValue);
+        } catch (NotEncodableValueException | NotNormalizableValueException $e) {
+            throw new \RuntimeException(
+                sprintf('There was a problem serializing property %s with value "%s".', $propertyName->value, get_debug_type($propertyValue)),
+                1594842314,
+                $e
+            );
+        }
+
+        if ($serializedPropertyValue === null) {
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:89]: Mmm, I find this tricky because I could see edge cases were serialization might result in a NULL value without the user really being able to control that?

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php [hunk 11]
```diff
@@ -66,4 +66,9 @@ public function merge(self $other): self
     {
         return new self(array_merge($this->values, $other->values));
     }
+
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:81]: To me the name of the DTO suggests that is already contains only stuff that is writeable, couldn't / shouldn't we just handle this directly in hte constructor (and withValue) such that we discard null values?

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyScope.php [hunk 12]
```diff
@@ -58,6 +60,15 @@ public function resolveAffectedOrigins(
         };
     }
 
+    public static function tryFromDeclaration(NodeType $nodeType, PropertyName $propertyName): self
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyScope.php:63]: How is this related to the PR?

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/NodeModification.php [hunk 13]
```diff
@@ -100,9 +109,30 @@ private function handleSetSerializedNodeProperties(
                     $affectedOrigin,
                     $nodeAggregate->getCoverageByOccupant($affectedOrigin),
                     $propertyValues,
+                    PropertyNames::createEmpty()
+                );
+            }
+        }
+
+        $propertiesToUnsetByScope = $this->splitPropertiesToUnsetByScope($command->propertiesToUnset, $nodeType);
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/NodeModification.php:110]: I guess that answers my question above regarding the scope... :)

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ContentStreamForking/ForkContentStreamWithoutDimensions.feature (hunk 9 above)
Comment (mhsdesign) [reply Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ContentStreamForking/ForkContentStreamWithoutDimensions.feature:49]: jain. `NodePropertiesWereSet` event publishing is discouraged in our tests and thus we dont need to make it more nice.https://github.com/neos/neos-development-collection/issues/4336

File: Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php (hunk 10 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:89]: Yes i know. I briefly discussed this with bastian freely after lets see :D

The thing is `node->hasProperty` will become more hard to describe and understand if `node->getProperty` can return `null` values if the node _has_ the property. What do you say?

But in the end we can always loosen this ...

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php (hunk 11 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:81]: the idea was not to break the high level api, that we allow null values at this point and split it later. We _need_ the information as we access it at some point. Might as well add a `getUnsetPropertyNames` to make it more explicit.

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyScope.php (hunk 12 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyScope.php:63]: extracted method to deduplicate code in `splitByScope` and `splitPropertiesToUnsetByScope`

... but ha i see you already noticed :D

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php (hunk 11 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:81]: Mmm, I guess, it#s just a weird method name vs DTO name and might not be obvious to anyone that you need to "cast" it with this method in some places... But if we need it

File: Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php (hunk 10 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:89]: I think the tricky part is that getProperty wil need to be able to return null, think policies, lets say you have an entity that not everyone can see, this would be reflected in the property, and the unserialiser shouldn't be able to unserialize it for some viewers.

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php (hunk 11 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:81]: Only we have to take care of casting it. And that is the tradeoff we discussed in favour of not breaking our api. Otherwise we would need to add a new property in the `SetNodeProperties` command itself.

We _could_ of course add this odd logic and splitting right into the `SetNodeProperties` ... by doing it on `create` or break things harder?

File: Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php (hunk 10 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:89]: I think we should open up the discussion when this case happens. Being more restrictive here in the first place seems valid?
Having an serializer enforce policies is also weird ....

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:89]: Actually I’m sure we need to catch this here 100%

if a serializer failed at write/command time that is okay

But the code below in deserializePropertyValue might apply to your concerns where I still find it confusing … is possible in the existing system for example to entity privilege protect an image or any asset? If so we probably do need to talk about that …

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php (hunk 11 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:81]: Nah it's fine, I see what you mean

File: Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php (hunk 10 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:89]: > Having an serializer enforce policies is also weird

Not enforce, but like, an entity privilege would result in a referenced object being null if the user reading the node doesn't have permission to see it.

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php (hunk 11 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:81]: @bwaidelich see this, as you commented the same as me (I hadn't seen your comment). I would also love to have it differnetly but if it's really that complicated?

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:81]: Bastis comment was on a previous state, we both worked out this one ...

File: Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php (hunk 10 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:89]: Yeah okay, reverted the read side error throwing as this is not handleable by the user. The write side is fine imo and desired.

Discussion how this might affect our api and confuse users can be continued here: https://github.com/neos/neos-development-collection/issues/4911

**Review by kitsunet — APPROVED** [review]
Cool, fine by me now, shall I hit merge?

**Comment by kitsunet** [comment]
(I guess we agreed in the weekyl this is fine)

**Comment by mhsdesign** [comment]
Yes i even did some additional testing that the command payload is also correctly migrated and a workspace with migrated changes can be merged.

Test cases:

- unset node properties via set serialised properties
- copy nodes which have explicitly unset properties which will also be added to the command
- create node with initial unset property

Each of the scenarios worked when being published after switching the branch and applying the `migrateevents:migratePropertiesToUnset` migration.

**Comment by mhsdesign** [comment]
Neos.Demo followup https://github.com/neos/Neos.Demo/pull/191
